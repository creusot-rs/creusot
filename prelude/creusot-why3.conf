[prover_modifiers]
name = "Z3"
driver = "z3_nombqi_seq.drv"