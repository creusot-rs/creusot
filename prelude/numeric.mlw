module BoundedInt

  use int.Int

  type t

  constant min : int
  constant max : int

  function to_int (n:t) : int
  meta coercion function to_int
  meta "model_projection" function to_int

  val to_int (n:t) : int
    ensures { result = n }

  predicate in_bounds (n:int) = min <= n <= max

  axiom to_int_in_bounds: forall n:t. in_bounds n

  val of_int (n:int) : t
    requires { [@expl:integer overflow] in_bounds n }
    ensures  { result = n }

  val function add (a:t) (b:t) : t
    requires { [@expl:integer overflow] in_bounds (a + b) }
    ensures   { result = a + b }

  val function sub (a:t) (b:t) : t
    requires { [@expl:integer overflow] in_bounds (a - b) }
    ensures  { result = a - b }

  val function mul (a:t) (b:t) : t
    requires { [@expl:integer overflow] in_bounds (a * b) }
    ensures  { result = a * b }

  val function neg (a:t) : t
    requires { [@expl:integer overflow] in_bounds (- a) }
    ensures  { result = - a }

  axiom extensionality: forall x y: t. to_int x = to_int y -> x = y

  val function eq (a:t) (b:t) : bool
    ensures { result <-> a = b }
    ensures { to_int a = to_int b -> result }

  val function ne (a:t) (b:t) : bool
    ensures { result <-> a <> b }
    ensures { to_int a <> to_int b -> result }

  val function le (a:t) (b:t) : bool
    ensures  { result <-> to_int a <= to_int b }

  val function lt (a:t) (b:t) : bool
    ensures  { result <-> to_int a < to_int b }

  val function ge (a:t) (b:t) : bool
    ensures  { result <-> to_int a >= to_int b }

  val function gt (a:t) (b:t) : bool
    ensures  { result <-> to_int a > to_int b }

  use int.ComputerDivision

  val function div (a:t) (b:t) : t
    requires { [@expl:division by zero] b <> 0 }
    requires { [@expl:integer overflow] in_bounds (div a b) }
    ensures  { result = div a b }

  val function rem (a:t) (b:t) : t
    requires { [@expl:division by zero] b <> 0 }
    requires { [@expl:integer overflow] in_bounds (mod a b) }
    ensures  { result = mod a b }

end

module Unsigned

  use int.Int

  let constant min_unsigned : int = 0

  clone export BoundedInt with
    constant min = min_unsigned, axiom .

  constant zero_unsigned : t

  axiom zero_unsigned_is_zero : to_int zero_unsigned = 0

  constant radix : int

  axiom radix_def : radix = max+1

end