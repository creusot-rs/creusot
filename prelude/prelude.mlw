module Prelude
  use Ref
  use mach.int.Int
  use mach.int.Int32
  use mach.int.Int64
  use mach.int.UInt32
  use mach.int.UInt64
  use string.Char
  use floating_point.Single
  use floating_point.Double

  type usize = uint64
  type isize = int64
  type opaque_ptr

  type borrowed 'a = { current : 'a ; final : 'a; }
  let function ( *_ ) x = x.current
  let function ( ^_ ) x = x.final
  val borrow_mut (a : 'a) : borrowed 'a
  ensures { *result = a }
end
