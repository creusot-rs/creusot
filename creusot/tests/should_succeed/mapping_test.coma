
module MappingTest_T_Type
  use prelude.prelude.Int32
  
  use prelude.prelude.Int
  
  type t_t  =
    | C_T int32
  
  function any_l (_ : 'b) : 'a
  
  let rec t_t (input:t_t) (ret  (a:int32))= any
    [ good (a:int32)-> {C_T a = input} (! ret {a}) | bad (a:int32)-> {C_T a <> input} {false} any ]
    
  
  function t_a (self : t_t) : int32 =
    match self with
      | C_T a -> a
      end
end
module MappingTest_Incr
  let%span smapping_test0 = "../mapping_test.rs" 31 16 31 31
  
  let%span smapping_test1 = "../mapping_test.rs" 32 14 32 15
  
  let%span smapping_test2 = "../mapping_test.rs" 35 19 35 50
  
  let%span smapping_test3 = "../mapping_test.rs" 27 12 27 24
  
  let%span smapping_test4 = "../mapping_test.rs" 28 12 28 26
  
  let%span smapping_test5 = "../mapping_test.rs" 29 11 29 37
  
  let%span span6 = "../mapping_test.rs" 19 4 21 74
  
  let%span span7 = "../mapping_test.rs" 23 8 23 14
  
  let%span span8 = "../../../../creusot-contracts/src/model.rs" 108 8 108 31
  
  let%span span9 = "../../../../creusot-contracts/src/model.rs" 90 8 90 31
  
  let%span span10 = "../../../../creusot-contracts/src/snapshot.rs" 28 20 28 48
  
  let%span span11 = "../../../../creusot-contracts/src/resolve.rs" 26 20 26 34
  
  use prelude.prelude.Int
  
  use map.Map
  
  use MappingTest_T_Type as T'0
  
  use prelude.prelude.Int32
  
  use MappingTest_T_Type as MappingTest_T_Type
  
  use map.Map
  
  use prelude.prelude.Int
  
  function shallow_model'0 [#"../mapping_test.rs" 22 4 22 50] (self : T'0.t_t) : Map.map int int
  
  axiom shallow_model'0_spec : forall self : T'0.t_t . [%#span6] forall i : int . Map.get (shallow_model'0 self) i
  = (if 0 <= i /\ i < Int32.to_int (MappingTest_T_Type.t_a self) then 1 else 0)
  
  use prelude.prelude.Borrow
  
  function shallow_model'2 (self : borrowed (T'0.t_t)) : Map.map int int =
    [%#span8] shallow_model'0 ( * self)
  
  use prelude.prelude.Snapshot
  
  use prelude.prelude.Intrinsic
  
  use map.Map
  
  use prelude.prelude.Snapshot
  
  function shallow_model'3 (self : borrowed (T'0.t_t)) : Map.map int int =
    [%#span9] shallow_model'2 self
  
  function shallow_model'1 (self : Snapshot.snap_ty (borrowed (T'0.t_t))) : Map.map int int =
    [%#span10] shallow_model'3 (Snapshot.inner self)
  
  predicate resolve'0 (self : borrowed (T'0.t_t)) =
    [%#span11]  ^ self =  * self
  
  use prelude.prelude.Int32
  
  use prelude.prelude.Snapshot
  
  let rec incr (t:borrowed (T'0.t_t)) (return'  (ret:()))= {[%#smapping_test4] Int32.to_int (MappingTest_T_Type.t_a ( * t))
    < 1000}
    {[%#smapping_test3] 0 <= Int32.to_int (MappingTest_T_Type.t_a ( * t))}
    (! bb0
    [ bb0 = s0 [ s0 =  [ &old_t <- [%#smapping_test0] Snapshot.new t ] s1 | s1 = bb1 ] 
    | bb1 = s0
      [ s0 = T'0.t_t { * t}
          (fun (ra'0:int32) ->
            Int32.add {ra'0} {[%#smapping_test1] (1 : int32)}
              (fun (_ret':int32) ->
                T'0.t_t { * t} (fun (l_a'0:int32) ->  [ &t <- { t with current = T'0.C_T _ret' ; } ] s1)))
      | s1 = -{resolve'0 t}- s2
      | s2 = {[@expl:assertion] [%#smapping_test2] shallow_model'0 ( ^ t)
        = Map.set (shallow_model'1 old_t) (Int32.to_int (MappingTest_T_Type.t_a ( * Snapshot.inner old_t))) 1}
        s3
      | s3 = return' {_0} ]
       ]
    )
    [ & _0 : () = any_l () | & t : borrowed (T'0.t_t) = t | & old_t : Snapshot.snap_ty (borrowed (T'0.t_t)) = any_l () ]
    
    [ return' (result:())-> {[@expl:postcondition] [%#smapping_test5] shallow_model'0 ( ^ t)
      = Map.set (shallow_model'2 t) (Int32.to_int (MappingTest_T_Type.t_a ( * t))) 1}
      (! return' {result}) ]
    
end
module MappingTest_F
  let%span smapping_test0 = "../mapping_test.rs" 39 23 39 25
  
  let%span smapping_test1 = "../mapping_test.rs" 40 19 40 34
  
  let%span smapping_test2 = "../mapping_test.rs" 41 19 41 34
  
  let%span smapping_test3 = "../mapping_test.rs" 43 19 43 34
  
  let%span smapping_test4 = "../mapping_test.rs" 44 19 44 34
  
  let%span span5 = "../../../../creusot-contracts/src/resolve.rs" 26 20 26 34
  
  let%span span6 = "../mapping_test.rs" 19 4 21 74
  
  let%span span7 = "../mapping_test.rs" 23 8 23 14
  
  let%span span8 = "../../../../creusot-contracts/src/model.rs" 108 8 108 31
  
  let%span span9 = "../mapping_test.rs" 27 12 27 24
  
  let%span span10 = "../mapping_test.rs" 28 12 28 26
  
  let%span span11 = "../mapping_test.rs" 29 11 29 37
  
  use prelude.prelude.Intrinsic
  
  use prelude.prelude.Borrow
  
  use MappingTest_T_Type as T'0
  
  predicate resolve'0 (self : borrowed (T'0.t_t)) =
    [%#span5]  ^ self =  * self
  
  use prelude.prelude.Int
  
  use map.Map
  
  use map.Map
  
  use prelude.prelude.Int32
  
  use MappingTest_T_Type as MappingTest_T_Type
  
  use map.Map
  
  use prelude.prelude.Int
  
  function shallow_model'0 [#"../mapping_test.rs" 22 4 22 50] (self : T'0.t_t) : Map.map int int
  
  axiom shallow_model'0_spec : forall self : T'0.t_t . [%#span6] forall i : int . Map.get (shallow_model'0 self) i
  = (if 0 <= i /\ i < Int32.to_int (MappingTest_T_Type.t_a self) then 1 else 0)
  
  function shallow_model'1 (self : borrowed (T'0.t_t)) : Map.map int int =
    [%#span8] shallow_model'0 ( * self)
  
  let rec incr'0 (t:borrowed (T'0.t_t)) (return'  (ret:()))= {[@expl:precondition] [%#span10] Int32.to_int (MappingTest_T_Type.t_a ( * t))
    < 1000}
    {[@expl:precondition] [%#span9] 0 <= Int32.to_int (MappingTest_T_Type.t_a ( * t))}
    any
    [ return' (result:())-> {[%#span11] shallow_model'0 ( ^ t)
      = Map.set (shallow_model'1 t) (Int32.to_int (MappingTest_T_Type.t_a ( * t))) 1}
      (! return' {result}) ]
    
  
  use prelude.prelude.Int32
  
  let rec f (_1:()) (return'  (ret:()))= (! bb0
    [ bb0 = s0
      [ s0 =  [ &x <- T'0.C_T ([%#smapping_test0] (42 : int32)) ] s1
      | s1 = {[@expl:assertion] [%#smapping_test1] Map.get (shallow_model'0 x) 13 = 1} s2
      | s2 = {[@expl:assertion] [%#smapping_test2] Map.get (shallow_model'0 x) 42 = 0} s3
      | s3 = Borrow.borrow_mut <T'0.t_t> {x} (fun (_ret':borrowed (T'0.t_t)) ->  [ &_8 <- _ret' ]  [ &x <-  ^ _8 ] s4)
      | s4 = Borrow.borrow_final <T'0.t_t> { * _8} {Borrow.get_id _8}
          (fun (_ret':borrowed (T'0.t_t)) ->  [ &_7 <- _ret' ]  [ &_8 <- { _8 with current = ( ^ _7) ; } ] s5)
      | s5 = incr'0 {_7} (fun (_ret':()) ->  [ &_6 <- _ret' ] s6)
      | s6 = bb1 ]
      
    | bb1 = s0
      [ s0 = -{resolve'0 _8}- s1
      | s1 = {[@expl:assertion] [%#smapping_test3] Map.get (shallow_model'0 x) 13 = 1} s2
      | s2 = {[@expl:assertion] [%#smapping_test4] Map.get (shallow_model'0 x) 42 = 1} s3
      | s3 = return' {_0} ]
       ]
    )
    [ & _0 : () = any_l ()
    | & x : T'0.t_t = any_l ()
    | & _6 : () = any_l ()
    | & _7 : borrowed (T'0.t_t) = any_l ()
    | & _8 : borrowed (T'0.t_t) = any_l () ]
     [ return' (result:())-> (! return' {result}) ] 
end
module MappingTest_Impl0
  
end
