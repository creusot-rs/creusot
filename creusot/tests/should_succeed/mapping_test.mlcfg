
module MappingTest_T_Type
  use prelude.Int32
  use prelude.Int
  type t_t  =
    | C_T int32
    
  let function t_a (self : t_t) : int32 = [@vc:do_not_keep_trace] [@vc:sp]
    match self with
      | C_T a -> a
      end
end
module MappingTest_Incr
  use prelude.Int
  use map.Map
  use prelude.Int32
  use MappingTest_T_Type as MappingTest_T_Type
  use map.Map
  use prelude.Int
  function shallow_model0 [#"../mapping_test.rs" 22 4 22 50] (self : MappingTest_T_Type.t_t) : Map.map int int
  val shallow_model0 [#"../mapping_test.rs" 22 4 22 50] (self : MappingTest_T_Type.t_t) : Map.map int int
    ensures { result = shallow_model0 self }
    
  axiom shallow_model0_spec : forall self : MappingTest_T_Type.t_t . [#"../mapping_test.rs" 19 4 21 74] forall i : int . Map.get (shallow_model0 self) i = (if 0 <= i /\ i < Int32.to_int (MappingTest_T_Type.t_a self) then
    1
  else
    0
  )
  use prelude.Borrow
  function shallow_model3 (self : borrowed (MappingTest_T_Type.t_t)) : Map.map int int =
    [#"../../../../creusot-contracts/src/model.rs" 97 8 97 31] shallow_model0 ( * self)
  val shallow_model3 (self : borrowed (MappingTest_T_Type.t_t)) : Map.map int int
    ensures { result = shallow_model3 self }
    
  use prelude.Snapshot
  use map.Map
  use prelude.Snapshot
  function shallow_model4 (self : borrowed (MappingTest_T_Type.t_t)) : Map.map int int =
    [#"../../../../creusot-contracts/src/model.rs" 79 8 79 31] shallow_model3 self
  val shallow_model4 (self : borrowed (MappingTest_T_Type.t_t)) : Map.map int int
    ensures { result = shallow_model4 self }
    
  function shallow_model1 (self : Snapshot.snap_ty (borrowed (MappingTest_T_Type.t_t))) : Map.map int int =
    [#"../../../../creusot-contracts/src/snapshot.rs" 27 20 27 48] shallow_model4 (Snapshot.inner self)
  val shallow_model1 (self : Snapshot.snap_ty (borrowed (MappingTest_T_Type.t_t))) : Map.map int int
    ensures { result = shallow_model1 self }
    
  predicate resolve0 (self : borrowed (MappingTest_T_Type.t_t)) =
    [#"../../../../creusot-contracts/src/resolve.rs" 26 20 26 34]  ^ self =  * self
  val resolve0 (self : borrowed (MappingTest_T_Type.t_t)) : bool
    ensures { result = resolve0 self }
    
  use prelude.Int32
  use prelude.Snapshot
  let rec cfg incr [#"../mapping_test.rs" 30 0 30 18] [@cfg:stackify] [@cfg:subregion_analysis] (t : borrowed (MappingTest_T_Type.t_t)) : ()
    requires {[#"../mapping_test.rs" 27 12 27 24] 0 <= Int32.to_int (MappingTest_T_Type.t_a ( * t))}
    requires {[#"../mapping_test.rs" 28 12 28 26] Int32.to_int (MappingTest_T_Type.t_a ( * t)) < 1000}
    ensures { [#"../mapping_test.rs" 29 11 29 37] shallow_model0 ( ^ t) = Map.set (shallow_model3 t) (Int32.to_int (MappingTest_T_Type.t_a ( * t))) 1 }
    
   = [@vc:do_not_keep_trace] [@vc:sp]
  var _0 : ();
  var t : borrowed (MappingTest_T_Type.t_t) = t;
  var old_t : Snapshot.snap_ty (borrowed (MappingTest_T_Type.t_t));
  {
    goto BB0
  }
  BB0 {
    [#"../mapping_test.rs" 31 16 31 31] old_t <- ([#"../mapping_test.rs" 31 16 31 31] Snapshot.new t);
    goto BB1
  }
  BB1 {
    [#"../mapping_test.rs" 32 4 32 15] t <- { t with current = (let MappingTest_T_Type.C_T x0 =  * t in MappingTest_T_Type.C_T (MappingTest_T_Type.t_a ( * t) + ([#"../mapping_test.rs" 32 14 32 15] (1 : int32)))) ; };
    assume { resolve0 t };
    assert { [@expl:assertion] [#"../mapping_test.rs" 35 19 35 50] shallow_model0 ( ^ t) = Map.set (shallow_model1 old_t) (Int32.to_int (MappingTest_T_Type.t_a ( * Snapshot.inner old_t))) 1 };
    [#"../mapping_test.rs" 30 19 36 1] _0 <- ([#"../mapping_test.rs" 30 19 36 1] ());
    return _0
  }
  
end
module MappingTest_F
  use prelude.Borrow
  use MappingTest_T_Type as MappingTest_T_Type
  predicate resolve0 (self : borrowed (MappingTest_T_Type.t_t)) =
    [#"../../../../creusot-contracts/src/resolve.rs" 26 20 26 34]  ^ self =  * self
  val resolve0 (self : borrowed (MappingTest_T_Type.t_t)) : bool
    ensures { result = resolve0 self }
    
  use prelude.Int
  use map.Map
  use map.Map
  use prelude.Int32
  use map.Map
  use prelude.Int
  function shallow_model0 [#"../mapping_test.rs" 22 4 22 50] (self : MappingTest_T_Type.t_t) : Map.map int int
  val shallow_model0 [#"../mapping_test.rs" 22 4 22 50] (self : MappingTest_T_Type.t_t) : Map.map int int
    ensures { result = shallow_model0 self }
    
  axiom shallow_model0_spec : forall self : MappingTest_T_Type.t_t . [#"../mapping_test.rs" 19 4 21 74] forall i : int . Map.get (shallow_model0 self) i = (if 0 <= i /\ i < Int32.to_int (MappingTest_T_Type.t_a self) then
    1
  else
    0
  )
  function shallow_model2 (self : borrowed (MappingTest_T_Type.t_t)) : Map.map int int =
    [#"../../../../creusot-contracts/src/model.rs" 97 8 97 31] shallow_model0 ( * self)
  val shallow_model2 (self : borrowed (MappingTest_T_Type.t_t)) : Map.map int int
    ensures { result = shallow_model2 self }
    
  val incr0 [#"../mapping_test.rs" 30 0 30 18] (t : borrowed (MappingTest_T_Type.t_t)) : ()
    requires {[#"../mapping_test.rs" 27 12 27 24] 0 <= Int32.to_int (MappingTest_T_Type.t_a ( * t))}
    requires {[#"../mapping_test.rs" 28 12 28 26] Int32.to_int (MappingTest_T_Type.t_a ( * t)) < 1000}
    ensures { [#"../mapping_test.rs" 29 11 29 37] shallow_model0 ( ^ t) = Map.set (shallow_model2 t) (Int32.to_int (MappingTest_T_Type.t_a ( * t))) 1 }
    
  use prelude.Int32
  let rec cfg f [#"../mapping_test.rs" 38 0 38 10] [@cfg:stackify] [@cfg:subregion_analysis] (_1 : ()) : ()
   = [@vc:do_not_keep_trace] [@vc:sp]
  var _0 : ();
  var x : MappingTest_T_Type.t_t;
  var _6 : ();
  var _7 : borrowed (MappingTest_T_Type.t_t);
  var _8 : borrowed (MappingTest_T_Type.t_t);
  {
    goto BB0
  }
  BB0 {
    [#"../mapping_test.rs" 39 16 39 27] x <- MappingTest_T_Type.C_T ([#"../mapping_test.rs" 39 23 39 25] (42 : int32));
    assert { [@expl:assertion] [#"../mapping_test.rs" 40 19 40 34] Map.get (shallow_model0 x) 13 = 1 };
    assert { [@expl:assertion] [#"../mapping_test.rs" 41 19 41 34] Map.get (shallow_model0 x) 42 = 0 };
    [#"../mapping_test.rs" 42 9 42 15] _8 <- Borrow.borrow_mut x;
    [#"../mapping_test.rs" 42 9 42 15] x <-  ^ _8;
    [#"../mapping_test.rs" 42 9 42 15] _7 <- Borrow.borrow_final ( * _8) (Borrow.get_id _8);
    [#"../mapping_test.rs" 42 9 42 15] _8 <- { _8 with current = ( ^ _7) ; };
    [#"../mapping_test.rs" 42 4 42 16] _6 <- ([#"../mapping_test.rs" 42 4 42 16] incr0 _7);
    _7 <- any borrowed (MappingTest_T_Type.t_t);
    goto BB1
  }
  BB1 {
    assume { resolve0 _8 };
    assert { [@expl:assertion] [#"../mapping_test.rs" 43 19 43 34] Map.get (shallow_model0 x) 13 = 1 };
    assert { [@expl:assertion] [#"../mapping_test.rs" 44 19 44 34] Map.get (shallow_model0 x) 42 = 1 };
    [#"../mapping_test.rs" 38 11 45 1] _0 <- ([#"../mapping_test.rs" 38 11 45 1] ());
    return _0
  }
  
end
module MappingTest_Impl0
  
end
