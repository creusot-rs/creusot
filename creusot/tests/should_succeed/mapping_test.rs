// basic tests for mappings

extern crate creusot_contracts;
use creusot_contracts::{
    logic::{Int, Mapping},
    *,
};

struct T {
    a: i32,
}

impl ShallowModel for T {
    type ShallowModelTy = Mapping<Int, Int>;

    #[ghost]
    #[open(self)]
    #[trusted]
    #[ensures(
        forall<i:Int>
            result.get(i) == (if 0 <= i && i < self.a@ { 1 } else { 0 }))]
    fn shallow_model(self) -> Self::ShallowModelTy {
        absurd
    }
}

#[requires( 0 <= (*t).a@ )] // otherwise its wrong !
#[requires( (*t).a@ < 1000 )] // to prevent overflow
#[ensures( (^t)@ == t@.set((*t).a@,1) )]
fn incr(t: &mut T) {
    let old_t = gh! { t };
    (*t).a += 1;
    // proving the post-consition via extensional equality of mappings
    // (notice `==` versus `==`)
    proof_assert!( (^t)@ == old_t@.set(old_t.a@,1) );
}

pub fn f() {
    let mut x = T { a: 42 };
    proof_assert!( x@.get(13) == 1 );
    proof_assert!( x@.get(42) == 0 );
    incr(&mut x);
    proof_assert!( x@.get(13) == 1 );
    proof_assert!( x@.get(42) == 1 );
}
