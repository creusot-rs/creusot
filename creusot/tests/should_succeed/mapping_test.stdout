module Type
  use Ref
  use mach.int.Int
  use prelude.Int8
  use prelude.Int16
  use mach.int.Int32
  use mach.int.Int64
  use prelude.UInt8
  use prelude.UInt16
  use mach.int.UInt32
  use mach.int.UInt64
  use string.Char
  use floating_point.Single
  use floating_point.Double
  use prelude.Prelude
  type mappingtest_t  = 
    | MappingTest_T int32
    
  function mappingtest_t_T_a (self : mappingtest_t) : int32
  val mappingtest_t_T_a (self : mappingtest_t) : int32
    ensures { result = mappingtest_t_T_a self }
    
  axiom mappingtest_t_T_a_acc : forall a : int32 . mappingtest_t_T_a (MappingTest_T a : mappingtest_t) = a
  type creusotcontracts_logic_ghost_ghost 't = 
    | CreusotContracts_Logic_Ghost_Ghost opaque_ptr
    
end
module CreusotContracts_Logic_Model_Model_ModelTy
  type self   
  type modelTy   
end
module CreusotContracts_Logic_Model_Model_Model_Interface
  type self   
  clone CreusotContracts_Logic_Model_Model_ModelTy as ModelTy0 with type self = self
  function model (self : self) : ModelTy0.modelTy
end
module CreusotContracts_Logic_Model_Model_Model
  type self   
  clone CreusotContracts_Logic_Model_Model_ModelTy as ModelTy0 with type self = self
  function model (self : self) : ModelTy0.modelTy
end
module MappingTest_Impl0_ModelTy
  use map.Map
  use mach.int.Int
  type modelTy  = 
    Map.map int int
end
module MappingTest_Impl0_Model_Interface
  use mach.int.Int
  use map.Map
  use mach.int.Int32
  use Type
  function model (self : Type.mappingtest_t) : Map.map int int
end
module MappingTest_Impl0_Model
  use mach.int.Int
  use map.Map
  use mach.int.Int32
  use Type
  function model (self : Type.mappingtest_t) : Map.map int int
  axiom model_spec : forall self : Type.mappingtest_t . forall i : (int) . Map.get (model self) i = (if 0 <= i && i < Int32.to_int (Type.mappingtest_t_T_a self) then
    1
  else
    0
  )
end
module MappingTest_Impl0
  use Type
  clone MappingTest_Impl0_Model as Model0 with axiom .
  clone MappingTest_Impl0_ModelTy as ModelTy0
  clone CreusotContracts_Logic_Model_Model_Model as Model1 with type self = Type.mappingtest_t,
  type ModelTy0.modelTy = ModelTy0.modelTy, function model = Model0.model
  clone CreusotContracts_Logic_Model_Model_ModelTy as ModelTy1 with type self = Type.mappingtest_t,
  type modelTy = ModelTy0.modelTy
end
module CreusotContracts_Logic_Resolve_Resolve_Resolve_Interface
  type self   
  predicate resolve (self : self)
end
module CreusotContracts_Logic_Resolve_Resolve_Resolve
  type self   
  predicate resolve (self : self)
end
module CreusotContracts_Logic_Ghost_Impl0_ModelTy
  type t   
  type modelTy  = 
    t
end
module CreusotContracts_Logic_Ghost_Impl0_Model_Interface
  type t   
  use Type
  function model (self : Type.creusotcontracts_logic_ghost_ghost t) : t
end
module CreusotContracts_Logic_Ghost_Impl0_Model
  type t   
  use Type
  function model (self : Type.creusotcontracts_logic_ghost_ghost t) : t
end
module CreusotContracts_Logic_Ghost_Impl0
  type t   
  use Type
  clone CreusotContracts_Logic_Ghost_Impl0_Model as Model0 with type t = t
  clone CreusotContracts_Logic_Ghost_Impl0_ModelTy as ModelTy0 with type t = t
  clone CreusotContracts_Logic_Model_Model_Model as Model1 with type self = Type.creusotcontracts_logic_ghost_ghost t,
  type ModelTy0.modelTy = ModelTy0.modelTy, function model = Model0.model
  clone CreusotContracts_Logic_Model_Model_ModelTy as ModelTy1 with type self = Type.creusotcontracts_logic_ghost_ghost t,
  type modelTy = ModelTy0.modelTy
end
module CreusotContracts_Logic_Ghost_Impl1_Record_Interface
  type t   
  use prelude.Prelude
  use Type
  clone CreusotContracts_Logic_Ghost_Impl0_Model_Interface as Model0 with type t = t
  val record [@cfg:stackify] (a : t) : Type.creusotcontracts_logic_ghost_ghost t
    ensures { Model0.model result = a }
    
end
module CreusotContracts_Logic_Ghost_Impl1_Record
  type t   
  use prelude.Prelude
  use Type
  clone CreusotContracts_Logic_Ghost_Impl0_Model_Interface as Model0 with type t = t
  val record [@cfg:stackify] (a : t) : Type.creusotcontracts_logic_ghost_ghost t
    ensures { Model0.model result = a }
    
end
module CreusotContracts_Logic_Resolve_Impl1_Resolve_Interface
  type t   
  use prelude.Prelude
  predicate resolve (self : borrowed t)
end
module CreusotContracts_Logic_Resolve_Impl1_Resolve
  type t   
  use prelude.Prelude
  predicate resolve (self : borrowed t) = 
     ^ self =  * self
end
module CreusotContracts_Logic_Resolve_Impl1
  type t   
  use prelude.Prelude
  clone CreusotContracts_Logic_Resolve_Impl1_Resolve as Resolve0 with type t = t
  clone CreusotContracts_Logic_Resolve_Resolve_Resolve as Resolve1 with type self = borrowed t,
  predicate resolve = Resolve0.resolve
end
module MappingTest_Incr_Interface
  use mach.int.Int32
  use mach.int.Int
  use map.Map
  use prelude.Prelude
  use Type
  clone MappingTest_Impl0_Model_Interface as Model0 with axiom .
  val incr [@cfg:stackify] (t : borrowed (Type.mappingtest_t)) : ()
    requires {Int32.to_int (Type.mappingtest_t_T_a ( * t)) < 1000}
    requires {0 <= Int32.to_int (Type.mappingtest_t_T_a ( * t))}
    ensures { Model0.model ( ^ t) = Map.set (Model0.model ( * t)) (Int32.to_int (Type.mappingtest_t_T_a ( * t))) 1 }
    
end
module MappingTest_Incr
  use mach.int.Int32
  use mach.int.Int
  use map.Map
  use map.MapExt
  use prelude.Prelude
  use Type
  clone CreusotContracts_Logic_Ghost_Impl0_Model as Model1 with type t = Type.mappingtest_t
  clone MappingTest_Impl0_Model as Model0 with axiom .
  clone CreusotContracts_Logic_Resolve_Resolve_Resolve as Resolve2 with type self = ()
  clone CreusotContracts_Logic_Resolve_Impl1_Resolve as Resolve1 with type t = Type.mappingtest_t
  clone CreusotContracts_Logic_Resolve_Resolve_Resolve as Resolve0 with type self = Type.creusotcontracts_logic_ghost_ghost (Type.mappingtest_t)
  clone CreusotContracts_Logic_Ghost_Impl1_Record_Interface as Record0 with type t = Type.mappingtest_t,
  function Model0.model = Model1.model
  let rec cfg incr [@cfg:stackify] (t : borrowed (Type.mappingtest_t)) : ()
    requires {Int32.to_int (Type.mappingtest_t_T_a ( * t)) < 1000}
    requires {0 <= Int32.to_int (Type.mappingtest_t_T_a ( * t))}
    ensures { Model0.model ( ^ t) = Map.set (Model0.model ( * t)) (Int32.to_int (Type.mappingtest_t_T_a ( * t))) 1 }
    
   = 
  var _0 : ();
  var t_1 : borrowed (Type.mappingtest_t);
  var old_t_2 : Type.creusotcontracts_logic_ghost_ghost (Type.mappingtest_t);
  var _3 : Type.mappingtest_t;
  var _4 : ();
  {
    t_1 <- t;
    goto BB0
  }
  BB0 {
    _3 <-  * t_1;
    old_t_2 <- Record0.record _3;
    goto BB1
  }
  BB1 {
    assume { Resolve0.resolve old_t_2 };
    t_1 <- { t_1 with current = (let Type.MappingTest_T a =  * t_1 in Type.MappingTest_T (Type.mappingtest_t_T_a ( * t_1) + (1 : int32))) };
    assume { Resolve1.resolve t_1 };
    assert { MapExt.(==) (Model0.model ( ^ t_1)) (Map.set (Model0.model (Model1.model old_t_2)) (Int32.to_int (Type.mappingtest_t_T_a (Model1.model old_t_2))) 1) };
    _4 <- ();
    assume { Resolve2.resolve _4 };
    _0 <- ();
    return _0
  }
  
end
module MappingTest_Main_Interface
  val main [@cfg:stackify] () : ()
end
module MappingTest_Main
  use mach.int.Int
  use mach.int.Int32
  use map.Map
  clone MappingTest_Impl0_Model as Model0 with axiom .
  use Type
  use prelude.Prelude
  clone CreusotContracts_Logic_Resolve_Impl1_Resolve as Resolve2 with type t = Type.mappingtest_t
  clone CreusotContracts_Logic_Resolve_Resolve_Resolve as Resolve1 with type self = Type.mappingtest_t
  clone CreusotContracts_Logic_Resolve_Resolve_Resolve as Resolve0 with type self = ()
  clone MappingTest_Incr_Interface as Incr0 with function Model0.model = Model0.model
  let rec cfg main [@cfg:stackify] () : () = 
  var _0 : ();
  var x_1 : Type.mappingtest_t;
  var _2 : ();
  var _3 : ();
  var _4 : ();
  var _5 : borrowed (Type.mappingtest_t);
  var _6 : borrowed (Type.mappingtest_t);
  var _7 : ();
  var _8 : ();
  {
    goto BB0
  }
  BB0 {
    x_1 <- Type.MappingTest_T (42 : int32);
    assert { Map.get (Model0.model x_1) 13 = 1 };
    _2 <- ();
    assume { Resolve0.resolve _2 };
    assert { Map.get (Model0.model x_1) 42 = 0 };
    _3 <- ();
    assume { Resolve0.resolve _3 };
    _6 <- borrow_mut x_1;
    x_1 <-  ^ _6;
    assume { Resolve1.resolve x_1 };
    _5 <- borrow_mut ( * _6);
    _6 <- { _6 with current = ( ^ _5) };
    _4 <- Incr0.incr _5;
    goto BB1
  }
  BB1 {
    assume { Resolve2.resolve _6 };
    assert { Map.get (Model0.model x_1) 13 = 1 };
    _7 <- ();
    assume { Resolve0.resolve _7 };
    assert { Map.get (Model0.model x_1) 42 = 1 };
    _8 <- ();
    assume { Resolve0.resolve _8 };
    _0 <- ();
    return _0
  }
  
end
